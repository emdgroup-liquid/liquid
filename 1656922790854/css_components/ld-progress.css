.ld-progress,:host{--ld-progress-height:var(--ld-sp-8);--ld-progress-steps-dot-width:var(--ld-progress-height);--ld-progress-valuemin:0;--ld-progress-valuemax:100;--ld-progress-valuenow:0;--ld-progress-transition-duration:0.2s;--ld-progress-width:17.5rem;--ld-progress-bar-col:var(--ld-thm-primary);--ld-progress-bar-col-overflow:var(--ld-thm-error);--ld-progress-bg-col:var(--ld-col-neutral-100);--ld-progress-bg-col-overflow:var(--ld-thm-error);--ld-progress-pending-gap-col:var(--ld-progress-bg-col);--ld-progress-pending-gap-col-overflow:var(--ld-thm-error-focus);--ld-progress-steps-col:var(--ld-col-neutral-200);--ld-progress-calc-valuenow:calc(var(--ld-progress-valuenow) - var(--ld-progress-valuemin));--ld-progress-calc-valuemax:calc(var(--ld-progress-valuemax) - var(--ld-progress-valuemin));--ld-progress-calc-relative-progress:calc(var(--ld-progress-calc-valuenow)/var(--ld-progress-calc-valuemax));--ld-progress-has-overflow:clamp(0,calc((var(--ld-progress-calc-relative-progress) - 1)*99999999999),1);border-radius:var(--ld-br-full);box-shadow:inset 0 0 0 calc((1 - var(--ld-progress-has-overflow))*99rem) var(--ld-progress-bg-col),inset 0 0 0 calc(var(--ld-progress-has-overflow)*99rem) var(--ld-col-wht-alpha-high),inset 0 0 0 calc(var(--ld-progress-has-overflow)*99rem) var(--ld-progress-bg-col-overflow);height:var(--ld-progress-height);max-width:100%;overflow:hidden;position:relative;width:var(--ld-progress-width)}.ld-progress:after,.ld-progress:before,:host:after,:host:before{content:"";display:block;height:100%;position:absolute;width:100%}.ld-progress:before,:host:before{background-color:var(--ld-progress-bar-col);opacity:calc(1 - var(--ld-progress-has-overflow));right:100%;transform:translateX(min(100%,calc(var(--ld-progress-calc-relative-progress)*100%)));transition:transform var(--ld-progress-transition-duration) ease,opacity var(--ld-progress-transition-duration) linear}.ld-progress:after,:host:after{background-color:var(--ld-progress-bar-col-overflow);left:100%;opacity:calc(var(--ld-progress-has-overflow));transform:translateX(max(-100%,calc((var(--ld-progress-calc-relative-progress) - 1)*-100%)));transition:transform var(--ld-progress-transition-duration) ease}.ld-progress--brand-color,:host(.ld-progress--brand-color){--ld-progress-bar-col:var(--ld-col-wht);--ld-progress-bar-col-overflow:var(--ld-thm-error);--ld-progress-bg-col:var(--ld-col-wht-alpha-medium);--ld-progress-bg-col-overflow:var(--ld-thm-error);--ld-progress-pending-gap-col:var(--ld-thm-primary-hover);--ld-progress-pending-gap-col-overflow:var(--ld-col-wht);--ld-progress-steps-col:var(--ld-col-wht-alpha-medium);box-shadow:inset 0 0 0 calc((1 - var(--ld-progress-has-overflow))*99rem) var(--ld-progress-bg-col),inset 0 0 0 calc(var(--ld-progress-has-overflow)*99rem) var(--ld-col-wht-alpha-medium),inset 0 calc(var(--ld-progress-has-overflow)*505rem) 500rem -500rem var(--ld-progress-bg-col-overflow)}.ld-progress--steps,:host(.ld-progress--steps){--ld-progress-step-gradient:var(--ld-progress-steps-col) 0%,var(--ld-progress-steps-col) calc(var(--ld-progress-steps-dot-width)/2),#0000 calc(var(--ld-progress-steps-dot-width)/2);background-blend-mode:multiply;background-image:radial-gradient(circle at left,var(--ld-progress-step-gradient)),radial-gradient(circle at right,var(--ld-progress-step-gradient));background-position:calc(var(--ld-progress-steps-dot-width)/2);background-repeat:repeat-x;background-size:calc((100% - var(--ld-progress-steps-dot-width) - 1px)/var(--ld-progress-calc-valuemax)) 100%;box-shadow:inset 0 0 0 calc(var(--ld-progress-has-overflow)*99rem) var(--ld-col-wht-alpha-lowest),inset 0 498rem calc(var(--ld-progress-has-overflow)*500rem) -500rem var(--ld-progress-bg-col-overflow)}.ld-progress--steps:after,.ld-progress--steps:before,:host(.ld-progress--steps):after,:host(.ld-progress--steps):before{border-radius:var(--ld-br-full)}.ld-progress--steps:before,:host(.ld-progress--steps):before{transform:translateX(min(100%,calc(var(--ld-progress-calc-relative-progress)*(100% - var(--ld-progress-steps-dot-width)) + var(--ld-progress-steps-dot-width))))}.ld-progress--steps:after,:host(.ld-progress--steps):after{transform:translateX(max(-100%,calc((var(--ld-progress-calc-relative-progress) - 1)*(100% - var(--ld-progress-steps-dot-width)) + var(--ld-progress-steps-dot-width) + 1px) * -1))}.ld-progress--steps.ld-progress--brand-color,:host(.ld-progress--steps.ld-progress--brand-color){box-shadow:inset 0 0 0 calc(var(--ld-progress-has-overflow)*99rem) var(--ld-col-wht-alpha-low),inset 0 500rem calc(var(--ld-progress-has-overflow)*500rem) -500rem var(--ld-progress-bg-col-overflow)}.ld-progress.ld-progress--indeterminate:before,:host(.ld-progress.ld-progress--indeterminate):before{transform:translateX(100%)}@keyframes ld-progress-animation-pending{0%{background-position-x:calc(-1.5rem + 1px)}to{background-position-x:0}}.ld-progress.ld-progress--pending:after,.ld-progress.ld-progress--pending:before,:host(.ld-progress.ld-progress--pending):after,:host(.ld-progress.ld-progress--pending):before{animation-duration:.5s;animation-iteration-count:infinite;animation-timing-function:linear;background-size:200% 100%}@media (prefers-reduced-motion:no-preference){.ld-progress.ld-progress--pending:after,.ld-progress.ld-progress--pending:before,:host(.ld-progress.ld-progress--pending):after,:host(.ld-progress.ld-progress--pending):before{animation-name:ld-progress-animation-pending}}.ld-progress.ld-progress--pending:before,:host(.ld-progress.ld-progress--pending):before{background-image:repeating-linear-gradient(-45deg,var(--ld-progress-pending-gap-col),var(--ld-progress-pending-gap-col) .5rem,var(--ld-progress-bar-col) .5rem,var(--ld-progress-bar-col) 1rem)}.ld-progress.ld-progress--pending:after,:host(.ld-progress.ld-progress--pending):after{animation-direction:reverse;background-image:repeating-linear-gradient(45deg,var(--ld-progress-pending-gap-col-overflow),var(--ld-progress-pending-gap-col-overflow) .5rem,var(--ld-progress-bar-col-overflow) .5rem,var(--ld-progress-bar-col-overflow) 1rem)}