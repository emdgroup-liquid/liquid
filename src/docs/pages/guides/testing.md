---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

