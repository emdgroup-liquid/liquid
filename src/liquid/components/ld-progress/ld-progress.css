:host,
.ld-progress {
  --ld-progress-height: var(--ld-sp-8);
  --ld-progress-steps-dot-width: var(--ld-progress-height);
  --ld-progress-valuemin: 0;
  --ld-progress-valuemax: 100;
  --ld-progress-valuenow: 0;
  --ld-progress-transition-duration: var(--ld-transition-duration-normal);
  --ld-progress-width: 17.5rem;

  /* colors */
  --ld-progress-bar-col: var(--ld-thm-primary);
  --ld-progress-bar-col-overflow: var(--ld-thm-error);
  --ld-progress-bg-col: var(--ld-col-neutral-100);
  --ld-progress-bg-col-overflow: var(--ld-thm-error);
  --ld-progress-pending-gap-col: var(--ld-progress-bg-col);
  --ld-progress-pending-gap-col-overflow: var(--ld-thm-error-focus);
  --ld-progress-steps-col: var(--ld-col-neutral-200);

  --ld-progress-calc-valuenow: calc(
    var(--ld-progress-valuenow) - var(--ld-progress-valuemin)
  );
  --ld-progress-calc-valuemax: calc(
    var(--ld-progress-valuemax) - var(--ld-progress-valuemin)
  );
  --ld-progress-calc-relative-progress: calc(
    var(--ld-progress-calc-valuenow) / var(--ld-progress-calc-valuemax)
  );
  --ld-progress-has-overflow: clamp(
    0,
    calc((var(--ld-progress-calc-relative-progress) - 1) * 99999999999),
    1
  );

  border-radius: var(--ld-br-full);
  box-shadow:
    inset 0 0 0 calc((1 - var(--ld-progress-has-overflow)) * 99rem)
      var(--ld-progress-bg-col),
    inset 0 0 0 calc(var(--ld-progress-has-overflow) * 99rem)
      var(--ld-col-wht-alpha-high),
    inset 0 0 0 calc(var(--ld-progress-has-overflow) * 99rem)
      var(--ld-progress-bg-col-overflow);
  display: block;
  height: var(--ld-progress-height);
  max-width: 100%;
  overflow: hidden;
  position: relative;
  width: var(--ld-progress-width);

  &::before,
  &::after {
    content: '';
    display: block;
    height: 100%;
    position: absolute;
    width: 100%;
  }

  &::before {
    background-color: var(--ld-progress-bar-col);
    opacity: calc(1 - var(--ld-progress-has-overflow));
    right: 100%;
    transform: translateX(
      min(100%, calc(var(--ld-progress-calc-relative-progress) * 100%))
    );
    transition:
      transform var(--ld-progress-transition-duration) ease,
      opacity var(--ld-progress-transition-duration) linear;
  }

  &::after {
    background-color: var(--ld-progress-bar-col-overflow);
    opacity: calc(var(--ld-progress-has-overflow));
    transform: translateX(
      max(-100%, calc((var(--ld-progress-calc-relative-progress) - 1) * -100%))
    ); /* - 1 substracts the "valid" progress value */
    transition: transform var(--ld-progress-transition-duration) ease;
    left: 100%;
  }
}

:host(.ld-progress--brand-color),
.ld-progress--brand-color {
  --ld-progress-bar-col: var(--ld-col-wht);
  --ld-progress-bar-col-overflow: var(--ld-thm-error);
  --ld-progress-bg-col: var(--ld-col-wht-alpha-medium);
  --ld-progress-bg-col-overflow: var(--ld-thm-error);
  --ld-progress-pending-gap-col: var(--ld-thm-primary-hover);
  --ld-progress-pending-gap-col-overflow: var(--ld-col-wht);
  --ld-progress-steps-col: var(--ld-col-wht-alpha-medium);

  box-shadow:
    inset 0 0 0 calc((1 - var(--ld-progress-has-overflow)) * 99rem)
      var(--ld-progress-bg-col),
    inset 0 0 0 calc(var(--ld-progress-has-overflow) * 99rem)
      var(--ld-col-wht-alpha-medium),
    inset 0 calc(var(--ld-progress-has-overflow) * 505rem) 500rem -500rem var(--ld-progress-bg-col-overflow);
}

:host(.ld-progress--steps),
.ld-progress--steps {
  --ld-progress-step-gradient: var(--ld-progress-steps-col) 0%,
    var(--ld-progress-steps-col) calc(var(--ld-progress-steps-dot-width) / 2),
    transparent calc(var(--ld-progress-steps-dot-width) / 2);
  background-image: radial-gradient(
      circle at left,
      var(--ld-progress-step-gradient)
    ),
    radial-gradient(circle at right, var(--ld-progress-step-gradient));
  background-blend-mode: multiply; /* removes gap in steps dot */
  background-size: calc(
      /* the 1px is a Safari HACK */
        (100% - var(--ld-progress-steps-dot-width) - 1px) /
        var(--ld-progress-calc-valuemax)
    )
    100%;
  background-repeat: repeat-x;
  background-position: calc(var(--ld-progress-steps-dot-width) / 2) center;
  box-shadow:
    inset 0 0 0 calc(var(--ld-progress-has-overflow) * 99rem)
      var(--ld-col-wht-alpha-lowest),
    inset 0 498rem calc(var(--ld-progress-has-overflow) * 500rem) -500rem var(--ld-progress-bg-col-overflow);

  &::before,
  &::after {
    border-radius: var(--ld-br-full);
  }

  &::before {
    transform: translateX(
      min(
        100%,
        calc(
          var(--ld-progress-calc-relative-progress) *
            (100% - var(--ld-progress-steps-dot-width)) +
            var(--ld-progress-steps-dot-width)
        )
      )
    );
  }

  &::after {
    transform: translateX(
      max(
        -100%,
        calc(
          (
              (var(--ld-progress-calc-relative-progress) - 1) *
                (100% - var(--ld-progress-steps-dot-width)) +
                /* the 1px is a Safari HACK */
                var(--ld-progress-steps-dot-width) + 1px
            ) * -1
        )
      )
    );
  }
}

:host(.ld-progress--steps.ld-progress--brand-color),
.ld-progress--steps.ld-progress--brand-color {
  box-shadow:
    inset 0 0 0 calc(var(--ld-progress-has-overflow) * 99rem)
      var(--ld-col-wht-alpha-low),
    inset 0 500rem calc(var(--ld-progress-has-overflow) * 500rem) -500rem var(--ld-progress-bg-col-overflow);
}

:host(.ld-progress.ld-progress--indeterminate),
.ld-progress.ld-progress--indeterminate {
  &::before {
    transform: translateX(100%);
  }
}

@keyframes ld-progress-animation-pending {
  from {
    background-position-x: calc(-1.5rem + 1px);
  }
  to {
    background-position-x: 0;
  }
}
:host(.ld-progress.ld-progress--pending),
.ld-progress.ld-progress--pending {
  &::before,
  &::after {
    background-size: 200% 100%;
    animation-duration: var(--ld-transition-duration-slow);
    animation-timing-function: linear;
    animation-iteration-count: infinite;

    @media (prefers-reduced-motion: no-preference) {
      animation-name: ld-progress-animation-pending;
      will-change: background-position-x;
    }
  }

  &::before {
    background-image: repeating-linear-gradient(
      -45deg,
      var(--ld-progress-pending-gap-col),
      var(--ld-progress-pending-gap-col) 0.5rem,
      var(--ld-progress-bar-col) 0.5rem,
      var(--ld-progress-bar-col) 1rem
    );
  }

  &::after {
    background-image: repeating-linear-gradient(
      45deg,
      var(--ld-progress-pending-gap-col-overflow),
      var(--ld-progress-pending-gap-col-overflow) 0.5rem,
      var(--ld-progress-bar-col-overflow) 0.5rem,
      var(--ld-progress-bar-col-overflow) 1rem
    );
    animation-direction: reverse;
  }
}
