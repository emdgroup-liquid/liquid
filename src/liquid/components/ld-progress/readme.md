---
eleventyNavigation:
  key: Progress
  parent: Components
layout: layout.njk
title: Progress
permalink: components/ld-progress/
---

<link rel="stylesheet" href="{{ env.base }}/{{ buildstamp }}css_components/ld-progress.css">
<link rel="stylesheet" href="{{ env.base }}/{{ buildstamp }}css_components/ld-sr-only.css">

# ld-progress

The `ld-progress` component can be used to displays the progress status for tasks that take a long time.

---

## Default

{% example %}
<ld-sr-only id="progress-label">Progress</ld-sr-only>
<ld-progress aria-labeledby="progress-label" aria-valuenow="25"></ld-progress>

<!-- React component -->

<LdSrOnly id="progress-label">Progress</LdSrOnly>
<LdProgress aria-labeledby="progress-label" aria-valuenow={25} />

<!-- CSS component -->

<span class="ld-sr-only" id="progress-label-css">Progress</span>
<div class="ld-progress"
     aria-labeledby="progress-label-css"
     aria-valuenow="25"
     role="progressbar"
     style="--ld-progress-valuenow: 25"></div>
{% endexample %}

Interactive example:

{% example '{ "stacked": true, "centered": true }' %}
<ld-sr-only id="progress-label">Progress</ld-sr-only>
<ld-progress aria-labeledby="progress-label" aria-valuenow="25"></ld-progress>

<ld-slider value="25" max="200" width="14rem"></ld-slider>

<script>
  void function() {
    const slider = document.currentScript.previousElementSibling
    const progress = slider.previousElementSibling

    slider.addEventListener('ldchange', ev => {
      progress.ariaValuenow = ev.detail[0]
    })
  }()
</script>

<!-- React component -->

const progressRef = useRef(null)

return (
  <>
    <LdSrOnly id="progress-label">Progress</LdSrOnly>
    <LdProgress
      aria-labeledby="progress-label"
      aria-valuenow={25}
      ref={progressRef}
     />

    <LdSlider
      value="25"
      max={200}
      width="14rem"
      onLdchange={(ev) => {
        progressRef.current.ariaValuenow = ev.detail[0]
      }}
    />
  </>
)

<!-- CSS component -->

<span class="ld-sr-only" id="progress-label-css">Progress</span>
<div class="ld-progress"
     aria-labeledby="progress-label-css"
     aria-valuenow="25"
     role="progressbar"
     style="--ld-progress-valuenow: 25"></div>

<ld-slider value="25" max="200" width="14rem"></ld-slider>

<script>
  void function() {
    const slider = document.currentScript.previousElementSibling
    const progress = slider.previousElementSibling
    slider.addEventListener('ldchange', ev => {
      const val = ev.detail[0]
      progress.ariaValuenow = val
      progress.style.setProperty('--ld-progress-valuenow', val)
    })
  }()
</script>
{% endexample %}

## With custom min and max values

{% example %}
<ld-progress aria-valuemax="300"
             aria-valuemin="100"
             aria-valuenow="150"></ld-progress>

<!-- React component -->

<LdProgress aria-valuemax={300}
            aria-valuemin={100}
            aria-valuenow={150} />

<!-- CSS component -->

<div class="ld-progress"
     aria-valuemax="300"
     aria-valuemin="100"
     aria-valuenow="150"
     role="progressbar"
     style="--ld-progress-valuemax: 300; --ld-progress-valuemin: 100; --ld-progress-valuenow: 150"></div>
{% endexample %}

## Overflow

The component can visualize an overflow value up to 200% of the maximum progress value.

{% example %}
<ld-progress aria-valuenow="125"></ld-progress>
<ld-progress aria-valuenow="225"></ld-progress>

<!-- React component -->

<LdProgress aria-valuenow={125} />
<LdProgress aria-valuenow={225} />

<!-- CSS component -->

<div class="ld-progress"
     aria-valuenow="125"
     role="progressbar"
     style="--ld-progress-valuenow: 125"></div>
<div class="ld-progress"
     aria-valuenow="225"
     role="progressbar"
     style="--ld-progress-valuenow: 225"></div>
{% endexample %}

## Steps

{% example %}
<ld-progress aria-valuemax="4" aria-valuenow="1" steps></ld-progress>

<!-- React component -->

<LdProgress aria-valuemax={4} aria-valuenow={1} steps />

<!-- CSS component -->

<div class="ld-progress ld-progress--steps"
     aria-valuemax="4"
     aria-valuenow="1"
     role="progressbar"
     style="--ld-progress-valuemax: 4; --ld-progress-valuenow: 1"></div>
{% endexample %}

## Pending

{% example %}
<ld-progress pending aria-valuetext="indeterminate"></ld-progress>
<ld-progress pending aria-valuenow="25"></ld-progress>

<!-- React component -->

<LdProgress pending aria-valuetext="indeterminate" />
<LdProgress pending aria-valuenow={25} />

<!-- CSS component -->

<div class="ld-progress ld-progress--pending"
     aria-valuetext="indeterminate"
     role="progressbar"
     style="--ld-progress-valuenow: 100"></div>
<div class="ld-progress ld-progress--pending"
     aria-valuenow="25"
     role="progressbar"
     style="--ld-progress-valuenow: 25"></div>
{% endexample %}

## On brand color

Use this mode on backgrounds with brand color.

{% example '{ "background": "brand", "hasBorder": false }' %}
<ld-progress brand-color aria-valuenow="25"></ld-progress>

<!-- React component -->

<LdProgress brandColor aria-valuenow={25} />

<!-- CSS component -->

<div class="ld-progress ld-progress--brand-color"
     aria-valuenow="25"
     role="progressbar"
     style="--ld-progress-valuenow: 25"></div>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description                                                                                                                                                                                | Type               | Default     |
| --------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------ | ----------- |
| `ariaLabeledby` | `aria-labeledby` | Defines the string value or identifies the element (or elements) that label the progressbar element providing an accessible name.                                                          | `string`           | `undefined` |
| `ariaValuemax`  | `aria-valuemax`  | Set to a decimal value representing the maximum value, and greater than aria-valuemin. If not present, the default value is 100.                                                           | `number`           | `100`       |
| `ariaValuemin`  | `aria-valuemin`  | Set to a decimal value representing the minimum value, and less than aria-valuemax. If not present, the default value is 0.                                                                | `number`           | `0`         |
| `ariaValuenow`  | `aria-valuenow`  | Only present and required if the value is not indeterminate. Set to a decimal value between 0, or valuemin if present, and aria-valuemax indicating the current value of the progress bar. | `number`           | `undefined` |
| `ariaValuetext` | `aria-valuetext` | Assistive technologies often present the value of aria-valuenow as a percentage. If this would not be accurate use this property to make the progress bar value understandable.            | `string`           | `undefined` |
| `brandColor`    | `brand-color`    | Styles the progress bar in a way that it looks good on the primary color of the current theme.                                                                                             | `boolean`          | `undefined` |
| `key`           | `key`            | for tracking the node's identity when working with lists                                                                                                                                   | `string \| number` | `undefined` |
| `pending`       | `pending`        | Used to show indeterminate or pending progress state.                                                                                                                                      | `boolean`          | `undefined` |
| `ref`           | `ref`            | reference to component                                                                                                                                                                     | `any`              | `undefined` |
| `steps`         | `steps`          | Devides progress bar in multiple progress steps.                                                                                                                                           | `boolean`          | `undefined` |


## Dependencies

### Used by

 - [ld-upload-item](../ld-file-upload/ld-upload-item)

### Graph
```mermaid
graph TD;
  ld-upload-item --> ld-progress
  style ld-progress fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
