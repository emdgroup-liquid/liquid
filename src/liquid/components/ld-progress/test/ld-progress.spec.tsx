import { newSpecPage } from '@stencil/core/testing'
import { LdProgress } from '../ld-progress'
import { h } from '@stencil/core'

describe('ld-progress', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdProgress],
      template: () => <ld-progress aria-valuenow="25" />,
    })
    expect(page.root).toMatchSnapshot()
  })

  describe('adds classes according to props', () => {
    it('adds indeterminate class', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => <ld-progress />,
      })
      expect(page.root).toHaveClass('ld-progress--indeterminate')
    })

    it('adds brand-color class', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => <ld-progress aria-valuenow="25" brand-color />,
      })
      expect(page.root).toHaveClass('ld-progress--brand-color')
    })

    it('adds pending class', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => <ld-progress aria-valuenow="25" pending />,
      })
      expect(page.root).toHaveClass('ld-progress--pending')
    })

    it('adds steps class', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => (
          <ld-progress aria-valuemax="5" aria-valuenow="2" steps />
        ),
      })
      expect(page.root).toHaveClass('ld-progress--steps')
    })
  })

  describe('applies css custom props according to props', () => {
    it('applies valuemax prop', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => <ld-progress aria-valuemax="200" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('applies valuemin prop', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => <ld-progress aria-valuemin="10" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('applies valuenow prop', async () => {
      const page = await newSpecPage({
        components: [LdProgress],
        template: () => <ld-progress aria-valuenow="25" />,
      })
      expect(page.root).toMatchSnapshot()
    })
  })
})
