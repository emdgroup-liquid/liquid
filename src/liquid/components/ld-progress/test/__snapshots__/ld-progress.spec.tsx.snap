// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-progress applies css custom props according to props applies valuemax prop 1`] = `
<ld-progress aria-valuemax="200" aria-valuemin="0" class="ld-progress ld-progress--indeterminate" role="progressbar" style="--ld-progress-valuemax: 200; --ld-progress-valuemin: 0;">
  <mock:shadow-root></mock:shadow-root>
</ld-progress>
`;

exports[`ld-progress applies css custom props according to props applies valuemin prop 1`] = `
<ld-progress aria-valuemax="100" aria-valuemin="10" class="ld-progress ld-progress--indeterminate" role="progressbar" style="--ld-progress-valuemax: 100; --ld-progress-valuemin: 10;">
  <mock:shadow-root></mock:shadow-root>
</ld-progress>
`;

exports[`ld-progress applies css custom props according to props applies valuenow prop 1`] = `
<ld-progress aria-valuemax="100" aria-valuemin="0" aria-valuenow="25" class="ld-progress" role="progressbar" style="--ld-progress-valuemax: 100; --ld-progress-valuemin: 0; --ld-progress-valuenow: 25;">
  <mock:shadow-root></mock:shadow-root>
</ld-progress>
`;

exports[`ld-progress renders 1`] = `
<ld-progress aria-valuemax="100" aria-valuemin="0" aria-valuenow="25" class="ld-progress" role="progressbar" style="--ld-progress-valuemax: 100; --ld-progress-valuemin: 0; --ld-progress-valuenow: 25;">
  <mock:shadow-root></mock:shadow-root>
</ld-progress>
`;
